### PR TITLE
talkpost do and preview too

### LINKED ISSUE #2551: Comment preview page: Use a normal reply form, maybe stop being a page
The comment preview page is one of the things rolled into talkpost_do.bml, and it sucks in several ways. Most notably:

- It uses its own hand-rolled version of the comment form instead of the talkform or the quickreply. 
- It shows the thread in reverse, with "comment context" positioned below your own comment and the form. 
- It uses a non-responsive site skin page. 
- It has that weird list of "allowed HTML" that seems to date back to the HTML3.2 days? What's up with that? I don't think I've seen it on any other page. 
- If you're logged in but commenting as a different site user, it displays you as anonymous during the preview and edit cycle, and doesn't expose the requested user's icons. (But it _does_ eventually post as the requested user.) 
    - Note that if logged in and commenting as an OpenID user, it _does_ display that user's proper name. But it still doesn't bring in their icons. 

Can we roll the comment preview functionality into S2 pages, the way comment edit currently is? That would provide a more realistic preview (especially useful if you're doing tables or anything that might get wacky in a journal's style), and would make the preview page less of a dumb special snowflake, so it could benefit from future improvements. Most of the functionality is already handled by `LJ::Talk::Post::make_preview`, so there might not even be _too_ much to extract from the BML?

### LINKED ISSUE #2572: Error loop when commenting as OpenID user while logged in as site user
Discovered while investigating #2542. 

- Make sure you're logged in as a site user. 
- Find a post with a captcha on it, and start replying to it using the full talkform.
- Use the radio buttons to comment as an OpenID user. Put in your ID url, but DON'T check the "log in" box. 
- Put in the wrong captcha response, so you'll get redirected back to talkpost_do to fix it.
- Enjoy. :) 

You'll end up with an error: `Invalid form submission. You may have left the reply form open too long, or logged out since you opened the page. Please try posting again.` If you keep trying to fix the captcha, enter your ID, and post, the page will mutate a bit, in a loop; note that the icon and username in the upper right of the page keeps switching between your openID user and your logged-in site user. You won't be able to comment as the openID user unless you select "log in". 
 
I think this predates anything I did to the talkform, but needs more investigation (with an old pre-talkform.tt version of the site) to be certain. 

I also think this is a rare enough behavior that no one has hit this bug in the wild, or at least no one has complained publicly about it.

### LINKED ISSUE #2646: Pull talkpost_do (reply error, preview) into s2foundation beta
I asked momiji for a favor because I'm mad at preview and don't feel like learning BML conversion yet. She got it into a controller, so now I gotta check for beta status and use that to switch between preview implementations. Talkform in reply errors oughtta be basically free.

### PR DESCRIPTION
**Pull request by nfagerlund**
MANY thanks to @rshatch for doing the BML conversion at the very start. 🙏🏼 I wasn't scared of the rest of this, but it was out on the far limits of my ability to understand this warty old codebase, and adding on one more thing to learn would have been too damn much. 

Fixes #2646 
Fixes #2551 
Fixes #2572 

This PR dynamites a ton of ancient garbage, including LJ::Talk::Post::init and LJ::Talk::Post::make_preview. A lot of DW's code seems to be working around the fact that no one wanted to touch these two functions ever again, so I bet this is gonna be a blast to review. But their replacements are fairly constrained and testable, as evinced by the fact that they HAVE tests now. 

Speaking of which, what are the cases you really want tested for `prepare_and_validate_comment`? I'm about at my limit for the authentication one, since logins act so screwy in testing, but I'd be down to add several more validation checks if people point out the cases that are most historically thorny. 

LMK if you want a tour through what's going on in here, but hopefully it's broken up into commit units that tell most of the story. The earlier commits with the really wild stuff (after the "scour BML from preview" one) are meant to always leave the site in a working state. 

After some thought, I decided that "check ban to lose the ability to ban" test actually serves as useful documentation, so I brought it in. 

<details><summary>Screens:</summary>

image

image

image
</details>

### CODE REVIEW

**Comment by nfagerlund** [comment]
GitHub is displaying my commits all out of order. Sigh.

**Comment by rahaeli** [comment]
It looks like you tried all the weird edge cases I would suggest, but please remind me when it's about to get pushed and I'll make sure to specifically call it out in the maint post about the push.

File: cgi-bin/LJ/Talk.pm [hunk 1]
```diff
@@ -2682,363 +2646,150 @@ sub init {
...
-        my $chal   = "$c_ditemid-$c_uid-$c_time-$c_chars";
-        my $secret = LJ::get_secret($c_time);
-        my $res    = Digest::MD5::md5_hex( $secret . $chal );
-        if ( $res ne $c_res ) {
-            $chrp_err = "invalid";
-        }
-        elsif ( $c_time < time() - 2 * 60 * 60 ) {
-            $chrp_err = "too_old" if $LJ::REQUIRE_TALKHASH_NOTOLD;
+        # No one's down with unvalidated site users.
+        # (FYI, nothing in -free or -nonfree ever calls that hook. -NF)
```
Comment (nfagerlund) [inline cgi-bin/LJ/Talk.pm:2735]:
```suggestion
        # (FYI, nothing in -free or -nonfree ever registers that hook. -NF)
```

File: cgi-bin/LJ/Talk.pm [hunk 2]
```diff
@@ -1612,25 +1612,17 @@ sub talkform {
...
-        %{ $opts->{styleopts} },
+        "replyto"      => $opts->{replyto},
+        "parenttalkid" => ( $opts->{replyto} + 0 ),
+        "itemid"       => $opts->{ditemid},
+        "journal"      => $journalu->{'user'},
+        "basepath"     => $basepath,
+        "dtid"         => $opts->{dtid},
+        "editid"       => $editid,
+        "chrp1"        => generate_chrp1($journalu->{userid}, $opts->{ditemid}),
+        %$styleopts,
```
Comment (nfagerlund) [inline cgi-bin/LJ/Talk.pm:1639]: (Sorry, ignore that uninitialized local variable; it's a glitch in the timestream from rebasing. That variable gets assigned during the "reply forms maintenance" commit.)

File: cgi-bin/LJ/Talk.pm [hunk 3]
```diff
@@ -2549,6 +2545,292 @@ sub enter_imported_comment {
...
+                "$tp_d.error.noopenidpost",
+                {
+                    aopts1 => "href='$LJ::SITEROOT/changeemail'",
+                    aopts2 => "href='$LJ::SITEROOT/register'"
+                }
+            );
+        }
+
+        # No one's down with unvalidated site users.
+        # (FYI, nothing in -free or -nonfree ever calls that hook. -NF)
```
Comment (zorkian) [inline cgi-bin/LJ/Talk.pm]: In the future, I've been ripping out any hook that we aren't calling. Simplifies reasoning about the code, and we can always re-add them in the future if needed. Most hooks were done to add ads and other "business options" with LJ, which isn't really applicable to us. Or other weird features.

File: cgi-bin/LJ/Talk.pm [hunk 4]
```diff
@@ -2549,6 +2545,292 @@ sub enter_imported_comment {
...
+        $parpost = LJ::Talk::get_talk2_row( $dbcr, $journalu->{userid}, $parenttalkid );
+        unless ($parpost) {
+            return $mlerr->("$tp_d.error.noparent");
+        }
+    }
+    # no replying to frozen comments
+    my $parent_state = $parpost->{state} // '';
+    return $mlerr->( "$tp_d.error.noreply_frozen" ) if $parent_state eq 'F';
+
+    # Next: Permissions checks! Easily solved, just become someone else.
```
Comment (zorkian) [inline cgi-bin/LJ/Talk.pm:2705]: lol

File: cgi-bin/LJ/Talk.pm [hunk 5]
```diff
@@ -2549,6 +2545,292 @@ sub enter_imported_comment {
...
+
+    # (For switching between variant error messages.)
+    my $iscomm = $journalu->is_community ? '.comm' : '';
+
+    if ( $commenter_is_user ) {
+
+        # test accounts can only comment on other test accounts.
+        if (   ( grep { $commenter->user eq $_ } @LJ::TESTACCTS )
```
Comment (zorkian) [inline cgi-bin/LJ/Talk.pm:2713]: Wow what is this functionality! Pretty sure we don't use TESTACCTS.

File: cgi-bin/LJ/Talk.pm (hunk 3 above)
Comment (kareila) [reply cgi-bin/LJ/Talk.pm]: Did we have an issue open at one point for looking at all the checks for hooks and ripping out all the ones that were never defined anywhere? Or was that just something we talked about doing?

File: cgi-bin/LJ/Talk.pm (hunk 5 above)
Comment (kareila) [reply cgi-bin/LJ/Talk.pm:2713]: I've never seen it before but it's referenced in a few other places too.

File: cgi-bin/LJ/Talk.pm [hunk 6]
```diff
@@ -2831,6 +2831,270 @@ sub prepare_and_validate_comment {
...
+# refactor this to just return an error instead; we only ever push one error,
+# since any error here is fatal.
+# - $didlogin: an optional scalar ref to indicate whether we started a new login
+# session. Only really used for an informational thing on an interstitial page
+# after commenting, which doesn't appear 100% reliably.
+# Returns one of:
+# - user object (authenticated user)
+# - undef (anon)
+# - string with URL (openid first pass)
+sub authenticate_user_and_mutate_form {
```
Comment (zorkian) [inline cgi-bin/LJ/Talk.pm]: I appreciate the descriptive names.

File: cgi-bin/LJ/Talk.pm (hunk 3 above)
Comment (nfagerlund) [reply cgi-bin/LJ/Talk.pm]: I think it just went past in conversation. Anyway, I'm 💯  for amending this PR to remove any more old garbage you happen to spot!

File: cgi-bin/LJ/Talk.pm (hunk 5 above)
Comment (nfagerlund) [reply cgi-bin/LJ/Talk.pm:2713]: yeah..... for returning a message that doesn't seem to exist from LJ::Protocol's login method, and for complaining at you if you try to change the test account's password in DW::Controller::Settings::changepassword_handler. Seems, uh,

File: cgi-bin/LJ/Talk.pm [hunk 7]
```diff
@@ -2831,6 +2831,270 @@ sub prepare_and_validate_comment {
...
+            my $iscomm = $journalu->is_community ? '.comm' : '';
+            my $noanon = $journalu->prop('opt_whocanreply') eq 'all' ? '' : '.noanon';
+            return $mlerr->( "$SC.error.nousername$noanon$iscomm", { sitename => $LJ::SITENAMESHORT } );
+        }
+
+        my $exptype;    # set to long if ! after username
+        my $ipfixed;    # set to remote  ip if < after username
+
+        # Parse inline login options.
+        # MUTATE FORM: remove trailing garbage from username.
```
Comment (zorkian) [inline cgi-bin/LJ/Talk.pm]: I really wonder if anybody actually uses this and would notice if it...magically stopped working. Besides me, anyway. Well, no reason to poke that horse I suppose.

File: cgi-bin/LJ/Talk.pm (hunk 2 above)
Comment (zorkian) [reply cgi-bin/LJ/Talk.pm:1639]: - [ ] 👍

File: cgi-bin/DW/Controller/Talk.pm [hunk 8]
```diff
@@ -75,38 +85,66 @@ sub talkpost_do_handler {
...
 
+    # validate the challenge/response value (anti-spammer)
+    my ($chrp_ok, $chrp_err) = LJ::Talk::validate_chrp1($POST->{'chrp1'});
+    unless ($chrp_ok) {
+        if ( $LJ::DEBUG{'talkspam'} ) {
+            my $ip = LJ::get_remote_ip();
+            my $ruser = $remote ? $remote->{user} : "[nonuser]";
+            carp( "talkhash error: from $ruser \@ $ip - $chrp_err - $talkurl\n" );
+        }
+        if ($LJ::REQUIRE_TALKHASH) {
```
Comment (zorkian) [inline cgi-bin/DW/Controller/Talk.pm:119]: I also have been baking in most of these tunables to match production. Unless we need it for development, might as well simplify conditionals and remove a lot of these.

File: cgi-bin/LJ/Talk.pm (hunk 3 above)
Comment (zorkian) [reply cgi-bin/LJ/Talk.pm]: @kareila don't think there's an issue open that I'm aware of, I've just been doing it from time to time in the name of code machete.

File: cgi-bin/LJ/Talk.pm (hunk 5 above)
Comment (zorkian) [reply cgi-bin/LJ/Talk.pm:2713]: Totally OK!

File: cgi-bin/LJ/Talk.pm (hunk 6 above)
Comment (nfagerlund) [reply cgi-bin/LJ/Talk.pm]: Thanks, I tried. (Partly I was just still mad about the vacuousness of "init", lol)

File: cgi-bin/LJ/Talk.pm (hunk 5 above)
Comment (rahaeli) [reply cgi-bin/LJ/Talk.pm:2713]: I feel like it's some kind of relic of LJ:IS_DEV_SERVER somehow?

File: cgi-bin/LJ/Talk.pm (hunk 7 above)
Comment (rahaeli) [reply cgi-bin/LJ/Talk.pm]: Put a counter on it, see how many ticks it tickies?

**Comment by zorkian** [comment]
I made it through each commit, and I have to say, I didn't find anything that I felt was substantive as far as review comments goes, which almost feels a shame, because of the amount of work you put in ... but it was well thought out, quality work and you even added tests.

So, tipping my hat to you, and punching that merge button.

**Comment by zorkian** [comment]
ALSO lest I forgot, to @momijizukamori whose enablement helped this go. Thank you!!!

**Comment by kareila** [comment]
@robodw-issues you think you are so helpful but no.

**Pull request merged**